### PR TITLE
Extract NavigationEngine processing logic and add tests

### PR DESCRIPTION
**Pull request by danesfeder**
Extracted logic from `NavigationEngine` into `NavigationRouteProcessor` and added tests.

Also fixes a bug where an `IndexOutOfBounds` was thrown when decoding the last step of a route.  Regression from #350.

Still TODO:
- [x] Finish tests for `NavigationRouteProcessor`

cc @akitchen for testing 👀

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+   */
+  private List<Point> decodeStepPoints(DirectionsRoute directionsRoute, int legIndex, int stepIndex) {
+    // Check for valid legs
+    List<RouteLeg> legs = directionsRoute.legs();
+    if (legs == null || legs.isEmpty()) {
+      return stepPoints;
+    }
+    // Check for valid steps
+    List<LegStep> steps = legs.get(legIndex).steps();
+    if (steps == null || steps.isEmpty() || stepIndex >= steps.size()) {
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java]: Actual bug fix is here `stepIndex >= steps.size()`

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationEngine.java [hunk 2]
```diff
@@ -93,19 +67,20 @@ private void handleRequest(final NewLocationModel newLocationModel) {
     final Location rawLocation = newLocationModel.location();
 
     // Generate a new route progress given the raw location update
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationEngine.java]: Could we remove the unnecessary comments?

> The best comment is a good name for a method or class.

👀 Code Smells 👉 [Comments](https://refactoring.guru/smells/comments)

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+import static com.mapbox.core.constants.Constants.PRECISION_6;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.checkBearingForStepCompletion;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.getSnappedLocation;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.increaseIndex;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.legDistanceRemaining;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.routeDistanceRemaining;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.stepDistanceRemaining;
+import static com.mapbox.services.android.navigation.v5.navigation.NavigationHelper.userSnappedToRoutePosition;
+
+class NavigationRouteProcessor implements OffRouteCallback {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java]: Could we remove the unnecessary comments in `NavigationRouteProcessor`? ☝️

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java (hunk 1 above)
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java]: What about [decomposing conditional](https://refactoring.com/catalog/decomposeConditional.html) extracting that complicated expression into a well-named `private` method? Conditionals tend to get more and more complicated in their logic over time.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+   * data and {@link NavigationIndices} needs to be reset.
+   *
+   * @param mapboxNavigation to get the current route and off-route engine
+   */
+  private void checkNewRoute(MapboxNavigation mapboxNavigation) {
+    DirectionsRoute directionsRoute = mapboxNavigation.getRoute();
+    if (RouteUtils.isNewRoute(previousRouteProgress, directionsRoute)) {
+      // Decode the first steps geometry and hold onto the resulting Position objects till the users
+      // on the next step. Indices are both 0 since the user just started on the new route.
+      stepPoints = decodeStepPoints(directionsRoute, 0, 0);
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java]: Magic numbers.

File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessorTest.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+import org.robolectric.annotation.Config;
+
+import java.util.List;
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.assertNotNull;
+import static junit.framework.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+@RunWith(RobolectricTestRunner.class)
```
Comment (Guardiola31337) [inline libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessorTest.java]: Why are we using the `RobolectricTestRunner`?

File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessorTest.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    routeProcessor.onShouldIncreaseIndex();
+    routeProcessor.checkIncreaseStepIndex(navigation);
+    RouteProgress secondProgress = routeProcessor.buildNewRouteProgress(navigation, mock(Location.class));
+    int secondStepIndex = secondProgress.currentLegProgress().stepIndex();
+    assertTrue(currentStepIndex != secondStepIndex);
+  }
+
+  @Test
+  public void onSnapToRouteEnabledAndUserOnRoute_snappedLocationReturns() throws Exception {
+    // Decode step points for snap-to-route logic
```
Comment (Guardiola31337) [inline libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessorTest.java]: Unnecessary comment.

File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessorTest.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+    Location snappedLocation = routeProcessor.buildSnappedLocation(
+      navigation, snapEnabled, rawLocation, progress, userOffRoute
+    );
+
+    assertTrue(!rawLocation.equals(snappedLocation));
+  }
+
+  @Test
+  public void onSnapToRouteDisabledAndUserOnRoute_rawLocationReturns() throws Exception {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessorTest.java]: Normally is good to structure the tests following the triple A 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40 http://wiki.c2.com/?ArrangeActAssert
If the `A`rrange part gets too long we should create factory or builder methods.

This applies to other tests included in `NavigationRouteProcessorTest`.

File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessorTest.java (hunk 5 above)
Comment (danesfeder) [reply libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessorTest.java]: For mocking `Location`

**Pull request merged**